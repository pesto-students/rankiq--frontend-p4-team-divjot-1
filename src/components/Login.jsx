import { useEffect, useState } from 'react';
import Container from '@mui/material/Container';
import Box from '@mui/material/Box';
import Grid from '@mui/material/Grid';
import Typography from '@mui/material/Typography';
import { styled } from '@mui/material/styles';
import { useController, useForm } from 'react-hook-form';
import TextField from '@mui/material/TextField';
import Divider from '@mui/material/Divider';
import Button from '@mui/material/Button';
import Link from '@mui/material/Link';
import Alert from '@mui/material/Alert';
import { useDispatch, useSelector } from 'react-redux';
import { useNavigate, Link as RouterLink } from 'react-router-dom';
import { isEmpty } from 'lodash';
import { REGEX, ERROR_MESSAGE } from '../constants';
import { signInUser } from '../ducks/auth';
import { accessTokenSelector, authDataSelector } from '../selectors';

const StyledLink = styled(Link)({
  textDecoration: 'none',
});

const StyledBox = styled(Box)(({ theme }) => ({
  backgroundColor: theme.palette.box.main,
  boxShadow: '0 3px 6px rgba(0,0,0,0.16), 0 3px 6px rgba(0,0,0,0.23)',
  borderRadius: '1.3125rem',
}));

function Login() {
  const dispatch = useDispatch();
  const navigate = useNavigate();
  const [showError, setShowError] = useState(false);
  const { error, loading } = useSelector(authDataSelector);
  const accessToken = useSelector(accessTokenSelector);

  useEffect(() => {
    if (!isEmpty(accessToken)) {
      navigate('/dashboard');
    }
  }, [navigate, accessToken]);
  const { control, formState, reset, getValues } = useForm({
    mode: 'onBlur',
    defaultValues: { email: '', password: '' },
  });

  const {
    field: { ref, ...userNameProps },
    fieldState: { invalid, error: userNameError },
  } = useController({
    name: 'email',
    control,
    rules: {
      required: ERROR_MESSAGE.REQUIRED,
      pattern: { value: REGEX.EMAIL, message: ERROR_MESSAGE.EMAIL },
    },
  });

  const {
    field: { ref: passwordRef, ...passwordProps },
    fieldState: { invalid: passwordInvalid, error: passwordError },
  } = useController({
    name: 'password',
    control,
    rules: {
      required: ERROR_MESSAGE.REQUIRED,
      pattern: { value: REGEX.PASSWORD, message: ERROR_MESSAGE.PASSWORD },
    },
  });

  const handleReset = () => {
    setShowError(false);
    reset();
  };

  return (
    <Container maxWidth="sm">
      <StyledBox
        sx={{ padding: { xs: '2rem 1rem', sm: '2rem' } }}
        mt={{ xs: '5rem', sm: '9rem' }}
        mb={{ xs: '1rem', sm: '5rem' }}
      >
        <form>
          <Grid container spacing={4}>
            <Grid item xs={12}>
              <Typography variant="h4" textAlign="center">
                Login
              </Typography>
            </Grid>
            <Grid item xs={12}>
              <TextField
                {...userNameProps}
                inputRef={ref}
                error={invalid}
                label="Email Id"
                type="email"
                helperText={userNameError?.message}
                required
                fullWidth
              />
            </Grid>
            <Grid item xs={12}>
              <TextField
                {...passwordProps}
                inputRef={passwordRef}
                error={passwordInvalid}
                label="Password"
                helperText={passwordError?.message}
                required
                type="password"
                fullWidth
              />
            </Grid>
            <Grid item xs={6}>
              <Button
                size="large"
                variant="contained"
                disabled={!formState.isValid || loading}
                fullWidth
                type="submit"
                onClick={() => {
                  setShowError(true);
                  dispatch(signInUser(getValues()));
                }}
              >
                Login
              </Button>
            </Grid>

            <Grid item xs={6}>
              <Button
                size="large"
                variant="outlined"
                onClick={handleReset}
                fullWidth
              >
                Reset
              </Button>
            </Grid>
            {!isEmpty(error) && showError && (
              <Grid item xs={12}>
                <Alert severity="error">{error}</Alert>
              </Grid>
            )}
            <Grid item xs={12}>
              <Box
                sx={{
                  display: 'flex',
                  columnGap: '0.5rem',
                  justifyContent: 'center',
                }}
              >
                <Typography variant="body1" textAlign="center">
                  Not on RankIQ yet?
                </Typography>
                <StyledLink
                  component={RouterLink}
                  variant="body1"
                  color="primary.main"
                  to="/signup"
                >
                  signup
                </StyledLink>
              </Box>
            </Grid>

            <Grid item xs={12}>
              <Divider>OR</Divider>
            </Grid>
            <Grid item xs={12}>
              <Button
                size="large"
                variant="contained"
                fullWidth
                onClick={() => {
                  navigate('/dashboard');
                }}
              >
                Continue as Guest
              </Button>
            </Grid>
          </Grid>
        </form>
      </StyledBox>
    </Container>
  );
}

export default Login;
