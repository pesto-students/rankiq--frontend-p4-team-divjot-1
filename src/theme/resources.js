import enDefault from './l10n/en.json';
import hiDefault from './l10n/hi.json';

const resources = {
  en: {
    default: enDefault,
  },
  hi: {
    default: hiDefault,
  },
};

export default resources;
