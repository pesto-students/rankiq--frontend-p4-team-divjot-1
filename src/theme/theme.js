import { createTheme } from '@mui/material/styles';

const themeOptions = {
  palette: {
    type: 'light',
    primary: {
      main: '#3f51b5',
    },
    secondary: {
      main: '#ffffff',
    },
    text: { navFooter: '#FFFFFF' },
    box: { main: '#FAFAFA' },
  },
};

const theme = createTheme(themeOptions);

export default theme;
