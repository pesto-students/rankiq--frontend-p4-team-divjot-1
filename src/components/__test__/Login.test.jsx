import { beforeEach, describe, expect, it, vi } from 'vitest';
import { render, screen } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import { Provider } from 'react-redux';
import { ThemeProvider } from '@mui/material/styles';
import { BrowserRouter } from 'react-router-dom';
import configureStore from 'redux-mock-store';
import thunk from 'redux-thunk';
import Login from '../Login';
import theme from '../../theme/theme';

const mockStore = configureStore([thunk]);
const navigate = vi.fn();
vi.mock('react-router-dom', async () => {
  return {
    ...(await vi.importActual('react-router-dom')),
    useNavigate: () => navigate,
  };
});

describe('Login.jsx - when user is not logged in', () => {
  let tree;
  let store;
  const storeData = {
    auth: {
      userInfo: {},
      accessToken: 'testToken',
      loading: false,
    },
  };

  beforeEach(() => {
    store = mockStore(storeData);
    tree = render(
      <Provider store={store}>
        <ThemeProvider theme={theme}>
          <BrowserRouter>
            <Login />
          </BrowserRouter>
        </ThemeProvider>
      </Provider>
    );
  });

  it('should render without any errors', () => {
    const { asFragment } = tree;
    expect(asFragment()).toMatchSnapshot();
  });

  it('should disable login button if invalid value is entered', async () => {
    const user = userEvent.setup();
    const emailField = screen.getByTestId('ipemail');
    const emailIp = emailField.getElementsByTagName('input')[0];
    const passwordField = screen.getByTestId('ippassword');
    await user.click(emailIp);
    await user.keyboard('test');
    await user.click(passwordField);
    expect(screen.queryByText('errors.email')).toBeInTheDocument();
    expect(screen.queryByTestId('btnSubmit')).toBeDisabled();
  });

  it('should enable login button if values are valid', async () => {
    const user = userEvent.setup();
    const emailField = screen.getByTestId('ipemail');
    const emailIp = emailField.getElementsByTagName('input')[0];
    const passwordField = screen.getByTestId('ippassword');
    const passwordIp = passwordField.getElementsByTagName('input')[0];
    await user.click(emailIp);
    await user.keyboard('test@test.com');
    await user.click(passwordIp);
    await user.keyboard('rankiq@123');
    await user.keyboard('{Tab}');
    await user.click(screen.queryByTestId('btnSubmit'));
    expect(screen.queryByTestId('btnSubmit')).not.toBeDisabled();
  });

  it('should reset form when reset button is clicked', async () => {
    const user = userEvent.setup();
    const emailField = screen.getByTestId('ipemail');
    const emailIp = emailField.getElementsByTagName('input')[0];
    const passwordField = screen.getByTestId('ippassword');
    const passwordIp = passwordField.getElementsByTagName('input')[0];
    await user.click(emailIp);
    await user.keyboard('test@test.com');
    await user.click(passwordIp);
    await user.keyboard('rankiq@123');
    await user.keyboard('{Tab}');
    await user.click(screen.queryByTestId('btnreset'));
    expect(screen.queryByTestId('btnSubmit')).toBeDisabled();
  });

  it('should open reset password modal when forgot pwd link is clicked', async () => {
    const user = userEvent.setup();
    const resetpwdLink = screen.getByTestId('forgotpwd');
    await user.click(resetpwdLink);
    expect(screen.queryByTestId('btnResetPwd')).toBeInTheDocument();
  });

  it('should navigate to dashboard page when user clicks on continue as guest', async () => {
    const user = userEvent.setup();
    const guestBtn = screen.getByTestId('btnGuest');
    await user.click(guestBtn);
    expect(navigate).toHaveBeenLastCalledWith('/dashboard');
  });
});

describe('Login.jsx - when user is logged in', () => {
  let store;
  const storeData = {
    auth: {
      userInfo: { email: 'email', firstName: 'Test Name' },
      accessToken: 'testToken',
    },
  };
  beforeEach(() => {
    store = mockStore(storeData);
    render(
      <Provider store={store}>
        <ThemeProvider theme={theme}>
          <BrowserRouter>
            <Login />
          </BrowserRouter>
        </ThemeProvider>
      </Provider>
    );
  });

  it('should navigate dashboard page if user is logged in', () => {
    expect(navigate).toHaveBeenLastCalledWith('/dashboard');
  });
});
