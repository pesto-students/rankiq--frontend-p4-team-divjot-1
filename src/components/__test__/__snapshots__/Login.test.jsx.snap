// Vitest Snapshot v1

exports[`Login.jsx - when user is not logged in > should render without any errors 1`] = `
<DocumentFragment>
  <div
    class="MuiContainer-root MuiContainer-maxWidthSm css-cuefkz-MuiContainer-root"
  >
    <div
      class="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation1 MuiCard-root css-1or43t3-MuiPaper-root-MuiCard-root"
    >
      <form>
        <div
          class="MuiGrid-root MuiGrid-container MuiGrid-spacing-xs-4 css-480o17-MuiGrid-root"
        >
          <div
            class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12 css-1idn90j-MuiGrid-root"
          >
            <h4
              class="MuiTypography-root MuiTypography-h4 css-149xwtl-MuiTypography-root"
            >
              generic.login
            </h4>
          </div>
          <div
            class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12 css-1idn90j-MuiGrid-root"
          >
            <div
              class="MuiFormControl-root MuiFormControl-fullWidth MuiTextField-root css-wb57ya-MuiFormControl-root-MuiTextField-root"
              data-testid="ipemail"
            >
              <label
                class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-outlined MuiFormLabel-colorPrimary Mui-required MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-outlined css-1iuujq-MuiFormLabel-root-MuiInputLabel-root"
                data-shrink="false"
                for=":r0:"
                id=":r0:-label"
              >
                login.email
                <span
                  aria-hidden="true"
                  class="MuiFormLabel-asterisk MuiInputLabel-asterisk css-wgai2y-MuiFormLabel-asterisk"
                >
                   *
                </span>
              </label>
              <div
                class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-colorPrimary MuiInputBase-fullWidth MuiInputBase-formControl css-xurl45-MuiInputBase-root-MuiOutlinedInput-root"
              >
                <input
                  aria-invalid="false"
                  class="MuiInputBase-input MuiOutlinedInput-input css-1t8l2tu-MuiInputBase-input-MuiOutlinedInput-input"
                  id=":r0:"
                  name="email"
                  required=""
                  type="email"
                  value=""
                />
                <fieldset
                  aria-hidden="true"
                  class="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
                >
                  <legend
                    class="css-yjsfm1"
                  >
                    <span>
                      login.email *
                    </span>
                  </legend>
                </fieldset>
              </div>
            </div>
          </div>
          <div
            class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12 css-1idn90j-MuiGrid-root"
          >
            <div
              class="MuiFormControl-root MuiFormControl-fullWidth MuiTextField-root css-wb57ya-MuiFormControl-root-MuiTextField-root"
              data-testid="ippassword"
            >
              <label
                class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-outlined MuiFormLabel-colorPrimary Mui-required MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-outlined css-1iuujq-MuiFormLabel-root-MuiInputLabel-root"
                data-shrink="false"
                for=":r1:"
                id=":r1:-label"
              >
                login.password
                <span
                  aria-hidden="true"
                  class="MuiFormLabel-asterisk MuiInputLabel-asterisk css-wgai2y-MuiFormLabel-asterisk"
                >
                   *
                </span>
              </label>
              <div
                class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-colorPrimary MuiInputBase-fullWidth MuiInputBase-formControl css-xurl45-MuiInputBase-root-MuiOutlinedInput-root"
              >
                <input
                  aria-invalid="false"
                  class="MuiInputBase-input MuiOutlinedInput-input css-1t8l2tu-MuiInputBase-input-MuiOutlinedInput-input"
                  id=":r1:"
                  name="password"
                  required=""
                  type="password"
                  value=""
                />
                <fieldset
                  aria-hidden="true"
                  class="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
                >
                  <legend
                    class="css-yjsfm1"
                  >
                    <span>
                      login.password *
                    </span>
                  </legend>
                </fieldset>
              </div>
            </div>
          </div>
          <div
            class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-6 css-1osj8n2-MuiGrid-root"
          >
            <button
              class="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeLarge MuiButton-containedSizeLarge MuiButton-fullWidth Mui-disabled MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeLarge MuiButton-containedSizeLarge MuiButton-fullWidth css-fjh6yo-MuiButtonBase-root-MuiButton-root"
              data-testid="btnSubmit"
              disabled=""
              tabindex="-1"
              type="submit"
            >
              generic.login
            </button>
          </div>
          <div
            class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-6 css-1osj8n2-MuiGrid-root"
          >
            <button
              class="MuiButtonBase-root MuiButton-root MuiButton-outlined MuiButton-outlinedPrimary MuiButton-sizeLarge MuiButton-outlinedSizeLarge MuiButton-fullWidth MuiButton-root MuiButton-outlined MuiButton-outlinedPrimary MuiButton-sizeLarge MuiButton-outlinedSizeLarge MuiButton-fullWidth css-i8s7vf-MuiButtonBase-root-MuiButton-root"
              data-testid="btnreset"
              tabindex="0"
              type="button"
            >
              generic.reset
              <span
                class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
              />
            </button>
          </div>
          <div
            class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12 css-1idn90j-MuiGrid-root"
          >
            <div
              class="MuiBox-root css-fxj1eo"
            >
              <p
                class="MuiTypography-root MuiTypography-body1 css-1uwgr7b-MuiTypography-root"
              >
                login.notOnRankiq
              </p>
              <a
                class="MuiTypography-root MuiTypography-body1 MuiLink-root MuiLink-underlineAlways css-1eui8x3-MuiTypography-root-MuiLink-root"
                href="/signup"
              >
                login.signup
              </a>
              <p
                class="MuiTypography-root MuiTypography-body1 css-1uwgr7b-MuiTypography-root"
              >
                |
              </p>
              <a
                class="MuiTypography-root MuiTypography-body1 MuiLink-root MuiLink-underlineAlways css-1eui8x3-MuiTypography-root-MuiLink-root"
                data-testid="forgotpwd"
                href="/"
              >
                login.resetPassword
              </a>
            </div>
            <div
              class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12 css-1idn90j-MuiGrid-root"
            >
              <div
                class="MuiBox-root css-16q7dti"
              >
                <p
                  class="MuiTypography-root MuiTypography-body1 css-1uwgr7b-MuiTypography-root"
                >
                  login.notOnRankiq
                </p>
                <a
                  class="MuiTypography-root MuiTypography-body1 MuiLink-root MuiLink-underlineAlways css-1eui8x3-MuiTypography-root-MuiLink-root"
                  href="/signup"
                >
                  login.signup
                </a>
              </div>
            </div>
            <div
              class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12 css-1idn90j-MuiGrid-root"
            >
              <div
                class="MuiBox-root css-16q7dti"
              >
                <a
                  class="MuiTypography-root MuiTypography-body1 MuiLink-root MuiLink-underlineAlways css-1eui8x3-MuiTypography-root-MuiLink-root"
                  href="/"
                >
                  login.resetPassword
                </a>
              </div>
            </div>
          </div>
          <div
            class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12 css-1idn90j-MuiGrid-root"
          >
            <div
              class="MuiDivider-root MuiDivider-fullWidth MuiDivider-withChildren css-1ws8119-MuiDivider-root"
              role="separator"
            >
              <span
                class="MuiDivider-wrapper css-qywfm8-MuiDivider-wrapper"
              >
                login.or
              </span>
            </div>
          </div>
          <div
            class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12 css-1idn90j-MuiGrid-root"
          >
            <button
              class="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeLarge MuiButton-containedSizeLarge MuiButton-fullWidth MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeLarge MuiButton-containedSizeLarge MuiButton-fullWidth css-fjh6yo-MuiButtonBase-root-MuiButton-root"
              data-testid="btnGuest"
              tabindex="0"
              type="button"
            >
              login.guest
              <span
                class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
              />
            </button>
          </div>
        </div>
      </form>
    </div>
  </div>
</DocumentFragment>
`;
